### PR TITLE
Build: handle symlinks in a centralized way

### PR DESCRIPTION
**Pull request by humitos**
Simplifies symlinks hanlding to centralize the security logic in one place only (`safe_open`); reducing the complexity of this code, making it easier to maintain and more secure.

Besides, it improves the UX for users to understand why their build is failing. It clearly communicates the error using our internal pattern for known exception managed at `Celery.on_failure`.

These are the highlights 🔆  of this work:

- Always follow symlinks
- Always copy symlink as-is. Only resolve them when uploading to storage.
- Always check symlinks are inside DOCROOT and `base_path` if passed
- Hard fail the build if there are symlinks in the output that are outside
  DOCROOT
- Centralize all the symlink checks in `safe_open` (used to open configuration
  files while building and to open files for uploading them)
- Clearly communicate the error to the user
- Move upload to storage function inside `Celery.execute` handler to allow us
  raising `UnsupportedSymlinkFileError` and using `on_failure` to clearly
  communicate the error to the user

I'm having issues running the tests right now, but I'll come back to this PR to update them soon.

> **NOTE** this PR is done over `rel` branch, but we should switch if to `main` when the other commit gets merged here.

### CODE REVIEW

File: readthedocs/core/utils/filesystem.py [hunk 1]
```diff
@@ -127,21 +97,11 @@ def safe_copytree(from_dir, to_dir, *args, **kwargs):
...
     """
-    path = Path(path)
-    if path.is_symlink():
-        log.info(
-            "Not deleting directory because it's a symlink.",
-            path=str(path),
-            resolved_path=path.resolve(),
-        )
-        return None
-    _assert_path_is_inside_docroot(path)
```
Comment (stsewd) [inline readthedocs/core/utils/filesystem.py:146]: rmtree fails hard (OSError) when trying to delete a dir that points to a symlink, but `_assert_path_is_inside_docroot` is still useful as mentioned in my other comment.

File: readthedocs/projects/tasks/builds.py [hunk 2]
```diff
@@ -514,29 +515,13 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
...
     def on_success(self, retval, task_id, args, kwargs):
-        html = self.data.outcomes['html']
-        search = self.data.outcomes['search']
-        localmedia = self.data.outcomes['localmedia']
-        pdf = self.data.outcomes['pdf']
-        epub = self.data.outcomes['epub']
-
-        time_before_store_build_artifacts = timezone.now()
-        # Store build artifacts to storage (local or cloud storage)
-        self.store_build_artifacts(
```
Comment (stsewd) [inline readthedocs/projects/tasks/builds.py:525]: we should probably do this change in another PR, doesn't look related to symlinks

File: readthedocs/builds/storage.py [hunk 3]
```diff
@@ -87,14 +88,6 @@ def copy_directory(self, source, destination):
...
             sub_destination = self.join(destination, filepath.name)
 
-            # Don't follow symlinks when uploading to storage.
-            if filepath.is_symlink():
-                log.info(
-                    "Skipping symlink upload.",
-                    path_resolved=str(filepath.resolve()),
-                )
-                continue
-
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/builds/storage.py:97]: I remove this check from here to allow communicating this decision, "do not follow symlinks", to the user clearly. Otherwise, with this check we are just ignoring symlinks without telling this to the user, which lead to unexpected behavior from a user perspective.

File: readthedocs/projects/tasks/builds.py (hunk 2 above)
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:525]: We need to move this call inside `execute` since we cannot raise exceptions from `on_success` and be able to manage them inside `on_failure` since it's not called when the code is already inside `on_success`.

File: readthedocs/core/utils/filesystem.py [hunk 4]
```diff
@@ -100,21 +85,6 @@ def safe_copytree(from_dir, to_dir, *args, **kwargs):
...
-            from_dir=from_dir,
-            from_dir_resolved=from_dir.resolve(),
-            to_dir=to_dir,
-            to_dir_resolved=to_dir.resolve(),
-        )
-        return False
-
-    _assert_path_is_inside_docroot(from_dir)
-    _assert_path_is_inside_docroot(to_dir)
-
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/utils/filesystem.py:117]: >  copytree will follow the symlink when doing the copy of the initial directory (there is a test for this as well)

What is that initial directory you refer to here?

[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/utils/filesystem.py:117]: The directory itself that you are going to copy (`from_dir`)

**Comment by humitos** [comment]
I think there are some good ideas implemented on this PR. However, this code has been changing due to https://github.com/readthedocs/readthedocs.org/pull/9842 and https://github.com/readthedocs/readthedocs.org/pull/9888

We can come back to this once we have those settled down for a while.

**Comment by humitos** [comment]
I'm closing this PR. It was useful to show my proposal about symlinks. Besides, the part related to store build artifacts was already implemented and merged from another PR.
